File: extension/src/protocol.ts
// SPDX-License-Identifier: MIT

/** The kind of item being saved. */
export type SaveKind = "article" | "image" | "video" | "quote";

/** Metadata sent alongside the Markdown body in a save request. */
export interface SaveRequestMetadata {
  kind: SaveKind;
  url: string;
  canonical_url: string;
  /**
   * A direct standalone image URL or legacy video identity. Current browser
   * video imports omit this field and stream bytes separately; the host writes
   * their local asset reference after the upload finishes. Article and quote
   * saves omit it.
   */
  media_url?: string;
  title: string;
  author?: string;
  site?: string;
  /** First page-declared theme colour; core validates and normalizes it for storage. */
  theme_color?: string;
  lang?: string;
  excerpt?: string;
  word_count?: number;
  saved_at: string;
}

/** The wire protocol version shared with the native host. Version 4 adds the
 * bounded, backpressured video-import actions below. */
export const PROTOCOL_VERSION = 4;

/**
 * One image the extension captured from the page (reusing the browser's cache),
 * base64-encoded so it can ride inside the JSON save message. The host decodes
 * and writes these; it never downloads anything itself.
 */
export interface ImageData {
  /** Source lookup key: a Markdown URL or an explicit preview/favicon role URL. */
  url: string;
  /** The response's `Content-Type`, used by the host to pick a file extension. */
  content_type: string;
  /** Standard base64 of the raw image bytes. */
  data_base64: string;
}

/** Message sent from the extension to the native host. */
export interface SaveRequest {
  protocol_version: typeof PROTOCOL_VERSION;
  action: "save";
  metadata: SaveRequestMetadata;
  markdown: string;
  images: ImageData[];
  /** Source URL whose captured bytes should become metadata.preview_asset. */
  preview_url?: string;
  /** Source URL whose captured bytes should become metadata.favicon_asset. */
  favicon_url?: string;
}

/**
 * Save an HTTP(S) URL without extracting cleaned article content. Live page
 * metadata and preview/favicon bytes may still be captured; the native host
 * routes them through the core's lightweight-link path so a later article
 * capture can upgrade the same URL-derived reading in place.
 */
export interface SaveLinkRequest {
  protocol_version: typeof PROTOCOL_VERSION;
  action: "save_link";
  metadata: SaveRequestMetadata;
  images: ImageData[];
  preview_url?: string;
  favicon_url?: string;
}

/** Metadata retained for a browser-captured local video import. The selected
 * source URL is deliberately absent; the host writes a local media_url when
 * the upload finishes. */
export type VideoImportMetadata = Omit<SaveRequestMetadata, "kind" | "media_url"> & {
  kind: "video";
  media_url?: never;
};

export interface VideoImportBeginRequest {
  protocol_version: typeof PROTOCOL_VERSION;
  action: "video_import_begin";
  upload_id: string;
  metadata: VideoImportMetadata;
  content_type: string;
  expected_bytes?: number;
}

export interface VideoImportChunkRequest {
  protocol_version: typeof PROTOCOL_VERSION;
  action: "video_import_chunk";
  upload_id: string;
  /** Zero-based and strictly monotonic within one upload. */
  sequence: number;
  /** Standard base64; decoded bytes are capped at 256 KiB per message. */
  data_base64: string;
}

export interface VideoImportFinishRequest {
  protocol_version: typeof PROTOCOL_VERSION;
  action: "video_import_finish";
  upload_id: string;
}

export interface VideoImportAbortRequest {
  protocol_version: typeof PROTOCOL_VERSION;
  action: "video_import_abort";
  upload_id: string;
}

export type VideoImportRequest =
  | VideoImportBeginRequest
  | VideoImportChunkRequest
  | VideoImportFinishRequest
  | VideoImportAbortRequest;

export interface SaveResponseSuccess {
  protocol_version: typeof PROTOCOL_VERSION;
  ok: true;
  id: string;
  path: string;
}

export interface SaveResponseError {
  protocol_version: typeof PROTOCOL_VERSION;
  ok: false;
  error: string;
  message: string;
}

export type SaveResponse = SaveResponseSuccess | SaveResponseError;

/** Begin/chunk/abort acknowledge without a saved-reading path. Finish returns
 * the existing SaveResponse success shape. */
export interface VideoImportAckSuccess {
  protocol_version: typeof PROTOCOL_VERSION;
  ok: true;
}

export type VideoImportResponse = VideoImportAckSuccess | SaveResponse;

/** Message sent from the extension to check whether a URL is already saved. */
export interface CheckRequest {
  protocol_version: typeof PROTOCOL_VERSION;
  action: "check";
  url: string;
}

export interface CheckResponse {
  protocol_version: typeof PROTOCOL_VERSION;
  ok: true;
  saved: boolean;
  id?: string;
}

File: core/native-host/src/protocol.rs
// SPDX-License-Identifier: MIT

use serde::{Deserialize, Serialize};

use cuttings_core::ReadingKind;

/// The wire protocol version shared with the browser extension.
pub const PROTOCOL_VERSION: u32 = 4;

/// Incoming save request from the browser extension.
#[derive(Debug, Deserialize)]
pub struct SaveRequest {
    pub protocol_version: u32,
    pub action: String,
    pub metadata: RequestMetadata,
    pub markdown: String,
    /// Image bytes captured by the extension. The host writes these to disk; it
    /// never downloads anything itself.
    pub images: Vec<RequestImage>,
    #[serde(default)]
    pub preview_url: Option<String>,
    #[serde(default)]
    pub favicon_url: Option<String>,
}

/// Incoming lightweight-link request with metadata/assets read from the live page.
#[derive(Debug, Deserialize)]
pub struct SaveLinkRequest {
    pub protocol_version: u32,
    pub action: String,
    pub metadata: RequestMetadata,
    pub images: Vec<RequestImage>,
    #[serde(default)]
    pub preview_url: Option<String>,
    #[serde(default)]
    pub favicon_url: Option<String>,
}

/// One image, captured and base64-encoded by the extension.
#[derive(Debug, Deserialize)]
pub struct RequestImage {
    /// Source lookup key: a Markdown URL or an explicit preview/favicon role URL.
    pub url: String,
    /// The `Content-Type` the browser saw, used to pick a file extension.
    #[serde(default)]
    pub content_type: String,
    /// Standard base64 of the raw image bytes.
    pub data_base64: String,
}

#[derive(Debug, Deserialize)]
pub struct RequestMetadata {
    #[serde(default)]
    pub kind: ReadingKind,
    pub url: String,
    #[serde(default)]
    pub media_url: Option<String>,
    pub canonical_url: String,
    pub title: String,
    pub author: Option<String>,
    pub site: Option<String>,
    #[serde(default)]
    pub theme_color: Option<String>,
    pub lang: Option<String>,
    pub excerpt: Option<String>,
    pub word_count: Option<u32>,
    pub saved_at: String,
}

/// Incoming check request — asks whether a URL is already in the library.
#[derive(Debug, Deserialize)]
pub struct CheckRequest {
    pub protocol_version: u32,
    pub url: String,
}

/// Start one connection-scoped browser video upload.
#[derive(Debug, Deserialize)]
pub struct VideoImportBeginRequest {
    pub protocol_version: u32,
    pub action: String,
    pub upload_id: String,
    pub metadata: RequestMetadata,
    pub content_type: String,
    #[serde(default)]
    pub expected_bytes: Option<u64>,
}

/// Append one base64-encoded chunk to the active browser video upload.
#[derive(Debug, Deserialize)]
pub struct VideoImportChunkRequest {
    pub protocol_version: u32,
    pub action: String,
    pub upload_id: String,
    pub sequence: u64,
    pub data_base64: String,
}

/// Finish or abort the active browser video upload.
#[derive(Debug, Deserialize)]
pub struct VideoImportEndRequest {
    pub protocol_version: u32,
    pub action: String,
    pub upload_id: String,
}

/// Outgoing response to the browser extension (shared by save and check actions).
#[derive(Debug, Serialize)]
pub struct SaveResponse {
    pub protocol_version: u32,
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    /// Present only in responses to a `check` action.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub saved: Option<bool>,
}

impl SaveResponse {
    pub fn ack() -> Self {
        Self {
            protocol_version: PROTOCOL_VERSION,
            ok: true,
            id: None,
            path: None,
            error: None,
            message: None,
            saved: None,
        }
    }

    pub fn success(id: String, path: String) -> Self {
        Self {
            protocol_version: PROTOCOL_VERSION,
            ok: true,
            id: Some(id),
            path: Some(path),
            error: None,
            message: None,
            saved: None,
        }
    }

    pub fn error(code: &str, msg: &str) -> Self {
        Self {
            protocol_version: PROTOCOL_VERSION,
            ok: false,
            id: None,
            path: None,
            error: Some(code.to_string()),
            message: Some(msg.to_string()),
            saved: None,
        }
    }

    pub fn check(is_saved: bool, id: Option<String>) -> Self {
        Self {
            protocol_version: PROTOCOL_VERSION,
            ok: true,
            saved: Some(is_saved),
            id,
            path: None,
            error: None,
            message: None,
        }
    }
}
